src/v/cloud_storage/topic_manifest.cc

```diff
@@ -353,56 +353,96 @@ void topic_manifest::do_update(const topic_manifest_handler& handler) {
-              fmt::format,
-              "Failed to parse topic manifest: {} at offset {}",
-              rapidjson::GetParseError_En(e),
-              o));
+            rapidjson::ParseErrorCode e = reader.GetParseErrorCode();
+            size_t o = reader.GetErrorOffset();
+
+            if (_topic_config) {
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "Failed to parse topic manifest {}: {} at offset {}",
+                  get_manifest_path(),
+                  rapidjson::GetParseError_En(e),
+                  o));
+            } else {
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "Failed to parse topic manifest: {} at offset {}",
+                  rapidjson::GetParseError_En(e),
+                  o));
+            }
         }
+        break;
+    }
+    case manifest_format::serde:
+        // serde format is straightforward: the buffer is a
+        // cluster::topic_configuration and _rev is inside
+        // remote_topic_properties
+        _topic_config = serde::from_iobuf<cluster::topic_configuration>(
+          std::move(result));
+        _rev = _topic_config->properties.remote_topic_properties.value()
+                 .remote_revision;
+        _manifest_version = topic_manifest::serde_version;
+        break;
     }
+
     co_return;
 }
 
 ss::future<serialized_data_stream> topic_manifest::serialize() const {
-    iobuf serialized;
-    iobuf_ostreambuf obuf(serialized);
-    std::ostream os(&obuf);
-    serialize(os);
-    if (!os.good()) {
-        throw std::runtime_error(fmt_with_ctx(
-          fmt::format,
-          "could not serialize topic manifest {}",
-          get_manifest_path()));
+    if (!_topic_config.has_value()) {
+        ss::throw_with_backtrace<std::runtime_error>(
+          "_topic_config is not initialized");
+    }
+
+    if (_manifest_version == first_version) {
+        // serialize in json format
+        iobuf serialized;
+        iobuf_ostreambuf obuf(serialized);
+        std::ostream os(&obuf);
+        serialize(os);
+        if (!os.good()) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "could not serialize topic manifest {}",
+              get_manifest_path()));
+        }
+        size_t size_bytes = serialized.size_bytes();
+        co_return serialized_data_stream{
+          .stream = make_iobuf_input_stream(std::move(serialized)),
+          .size_bytes = size_bytes};
     }
-    size_t size_bytes = serialized.size_bytes();
+
+    // serialize in binary format
+    auto to_serialize = _topic_config.value();
+    // insert _rev into remote_topic_properties, to be able to extract it when
+    // deserializing it
+    to_serialize.properties.remote_topic_properties
+      = cluster::remote_topic_properties(_rev, to_serialize.partition_count);
+
+    auto serialized = serde::to_iobuf(std::move(to_serialize));
```

Comment:
Is this just a convenient location to store the revision, or is this actually where we _should_ be tracking the revision? If the latter, do we even need the `_rev` field anymore? If the former, I think we should consider making some `topic_manifest_state` serde::envelope type that includes just the topic config and revision, and make this `serialize()` call just call into the default serde methods. That way it feels easier to evolve the fields in the manifest if need be

Reply:
well the _rev field ends up populating remote_topic_properties in our codebase, when reading a recovery topic.
so putting it here directly feels the right place.

but i guess in general it doesn't hurt to have a topic_manifest_state: public serde type. just i don't think it should track revision separatedly. it would be useful if we want to persist something that is not a topic_configuration

Reply:
I see. Yeah it makes sense, it does seem like the right place to me too.

Wondering if it we can just remove `_rev` entirely then and rely on the topic properties. It seems confusing otherwise to keep them separate

Reply:
FWIW this is just a nit, since it doesn't change how we serialize. Just a thought about collapsing the `_rev` field in with what we have in remote properties

Reply:
My point about collapsing `_rev` into the remote topic properties is just a nit.

My question about `topic_manifest_state` could be even simpler though: would it make sense to make `topic_manifest` an envelope and have its `serde_fields` only include `topic_properties`? I don't feel too strongly about this, since as you say, the topic manifest is currently 1:1 with the topic config.

Reply:
done